src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+    /// assert_eq!(doc.get_i64("unknown"), Ok(None));
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn get_i64(&self, key: &str) -> RawResult<Option<i64>> {
+        self.get_with(key, elem::RawBson::as_i64)
+    }
+
+    /// Return a reference to the contained data as a `&[u8]`
+    ///
+    /// ```
+    /// # use bson::raw::RawDocument;
+    /// use bson::doc;
+    /// let docbuf = RawDocument::from_document(&doc!{});
+    /// assert_eq!(docbuf.as_bytes(), b"\x05\x00\x00\x00\x00");
+    /// ```
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+}
+
+impl AsRef<RawDocumentRef> for RawDocumentRef {
+    fn as_ref(&self) -> &RawDocumentRef {
+        self
+    }
+}
+
+impl Deref for RawDocument {
+    type Target = RawDocumentRef;
+
+    fn deref(&self) -> &Self::Target {
+        RawDocumentRef::new_unchecked(&self.data)
+    }
+}
+
+impl TryFrom<&RawDocumentRef> for crate::Document {
+    type Error = RawError;
+
+    fn try_from(rawdoc: &RawDocumentRef) -> RawResult<Document> {
+        rawdoc
+            .into_iter()
+            .map(|res| res.and_then(|(k, v)| Ok((k.to_owned(), v.try_into()?))))
+            .collect()
+    }
+}
+
+impl<'a> IntoIterator for &'a RawDocumentRef {
+    type IntoIter = RawDocumentIter<'a>;
+    type Item = RawResult<(&'a str, RawBson<'a>)>;
+
+    fn into_iter(self) -> RawDocumentIter<'a> {
+        RawDocumentIter {
+            doc: self,
+            offset: 4,
+        }
+    }
+}
+
+pub struct RawDocumentIter<'a> {
+    doc: &'a RawDocumentRef,
+    offset: usize,
+}
+
+impl<'a> Iterator for RawDocumentIter<'a> {
+    type Item = RawResult<(&'a str, elem::RawBson<'a>)>;
+
+    fn next(&mut self) -> Option<RawResult<(&'a str, elem::RawBson<'a>)>> {
+        if self.offset == self.doc.data.len() - 1 {
+            if self.doc.data[self.offset] == 0 {
+                // end of document marker
+                return None;
+            } else {
+                return Some(Err(RawError::MalformedValue(
+                    "document not null terminated".into(),
+                )));
+            }
+        }
+
+        let key = match read_nullterminated(&self.doc.data[self.offset + 1..]) {
+            Ok(key) => key,
+            Err(err) => return Some(Err(err)),
+        };
+
+        let valueoffset = self.offset + 1 + key.len() + 1; // type specifier + key + \0
+
+        let element_type = match ElementType::from(self.doc.data[self.offset]) {
+            Some(et) => et,
+            None => {
+                return Some(Err(RawError::MalformedValue(format!(
+                    "invalid tag: {}",
+                    self.doc.data[self.offset]
+                ))))
+            }
+        };
+
+        let element_size = match element_type {
+            ElementType::Double => 8,
+            ElementType::String => {
+                let size =
+                    4 + i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize;
```

Comment:
this may panic if the data is too short, since we never check that valueoffset is less than `data.len()`. This applies in a lot of other places in this function.